go/src/koding/kites/kloud/provider/google/machine.go

```diff
@@ -0,0 +1,204 @@
+package google
+
+import (
+	"errors"
+	"net/http"
+
+	"koding/kites/kloud/eventer"
+	"koding/kites/kloud/machinestate"
+	"koding/kites/kloud/stack"
+	"koding/kites/kloud/stack/provider"
+	"koding/kites/kloud/waitstate"
+
+	"golang.org/x/net/context"
+	compute "google.golang.org/api/compute/v1"
+	"google.golang.org/api/googleapi"
+)
+
+// Machine represents a single MongodDB document from the jMachines
+// collection.
+type Machine struct {
+	*provider.BaseMachine
+
+	InstancesService *compute.InstancesService
+}
+
+var (
+	_ provider.Machine = (*Machine)(nil) // public API
+	_ stack.Machiner   = (*Machine)(nil) // internal API
+)
+
+func newMachine(bm *provider.BaseMachine) (provider.Machine, error) {
+	m := &Machine{BaseMachine: bm}
+	cred := m.Cred()
+
+	computeService, err := cred.ComputeService()
+	if err != nil {
+		return nil, err
+	}
+
+	m.InstancesService = compute.NewInstancesService(computeService)
+	return m, nil
+}
+
+// Start starts Google compute instance.
+func (m *Machine) Start(ctx context.Context) (interface{}, error) {
+	ev, withPush := eventer.FromContext(ctx)
+	if withPush {
+		ev.Push(&eventer.Event{
+			Message:    "Starting machine",
+			Status:     machinestate.Starting,
+			Percentage: 25,
+		})
+	}
+
+	project, zone, name := m.Location()
+	_, err := m.InstancesService.Start(project, zone, name).Do()
+	// Ignore http.StatusNotModified status.
+	if err != nil && googleapi.IsNotModified(err) {
+		if withPush {
+			ev.Push(&eventer.Event{
+				Message:    "Machine is running",
+				Status:     machinestate.Running,
+				Percentage: 60,
+			})
+		}
+		return nil, nil
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	stateFunc := func(currentPercentage int) (machinestate.State, error) {
+		if withPush {
+			ev.Push(&eventer.Event{
+				Message:    "Starting machine",
+				Status:     machinestate.Starting,
+				Percentage: currentPercentage,
+			})
+		}
+
+		state, _, err := m.Info(nil)
+		return state, err
+	}
+
+	ws := waitstate.WaitState{
+		StateFunc:    stateFunc,
+		DesiredState: machinestate.Running,
+		Start:        45,
+		Finish:       60,
+	}
+
+	return nil, ws.Wait()
+}
+
+// Start stops Google compute instance.
+func (m *Machine) Stop(ctx context.Context) (interface{}, error) {
+	ev, withPush := eventer.FromContext(ctx)
```

Comment:
We could move this to BaseMachine to reduce further the provider boilerplate. Commenting for posterity.